benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,281 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <gflags/gflags.h>
+
+
+#ifdef HAS_CUDA
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include "cuda/base/device_guard.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+#ifdef HAS_HIP
+
+
+#include <hip/hip_runtime.h>
+
+
+#include "hip/base/device_guard.hip.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+DEFINE_bool(gpu_timer, false,
+            "use gpu timer based on event. It is valid only when "
+            "executor is cuda or hip");
+
+
+class Timer {
+public:
+    void tic()
+    {
+        assert(tic_called_ == false);
+        this->tic_impl();
+        tic_called_ = true;
+    }
+
+    std::size_t toc()
```

Comment:
Why is it nice to have the return type as `size_t`? I see that Cuda and Hip timers actually return `float` which are then cast to `size_t`. Maybe the return type can just be kept as `double`?

Reply:
It is based on the chrono library.
In chrono, we can count the number of nanosecond, so I use size_t to represent it.

Reply:
Okay. But it does not seem like good practice to cast a `float` to `size_t`... though the other way round (or better yet to `double`) is not too bad. [Side note: [cppreference](https://en.cppreference.com/w/cpp/chrono/duration) says `nanosecond` is a signed 64 bit type, not `size_t`].

Reply:
I don't think it is a good idea to have `toc()` return something because it is not obvious what it returns and in any case, the return value is not really useful. It could be any of these return values:
- The number of times the time has already been recorded
- the average runtime up until this point
- the sum of the runtime until this point
- the runtime of the latest measurement (you have chosen this one)

Additionally, it is not clear that the unit returned is `nanoseconds`. All in all, I would recommend to have this function not return anything.
If you want to know the latest result, I would prefer an additional function (`std::chrono` does the same).